graphql_compiler/schema/schema_info.py

```diff
@@ -20,20 +20,46 @@
 from ..schema_generation.schema_graph import SchemaGraph
 
 
-# Describes the intent to join two tables using the specified columns.
-#
-# The resulting join expression could be something like:
-# JOIN origin_table.from_column = destination_table.to_column
-#
-# The type of join (inner vs left, etc.) is not specified.
-# The tables are not specified.
-DirectJoinDescriptor = namedtuple(
-    "DirectJoinDescriptor",
-    (
-        "from_column",  # The column in the source table we intend to join on.
-        "to_column",  # The column in the destination table we intend to join on.
-    ),
-)
+@dataclass
+class DirectJoinDescriptor:
+    """Describes the ability to join two tables using the specified columns.
+
+    The resulting join expression could be something like:
+    JOIN origin_table.from_column = destination_table.to_column
+
+    The type of join (inner vs left, etc.) is not specified.
+    The tables are not specified.
+    """
+
+    from_column: str  # The column in the source table we intend to join on.
+    to_column: str  # The column in the destination table we intend to join on.
+
+
+@dataclass
+class CompositeJoinDescriptor:
+    """Describes the ability to join two tables with a composite relationship.
+
+    The resulting join expression could be something like:
+    JOIN
+        origin_table.from_column_1 == destination_table.to_column_1 AND
+        origin_table.from_column_2 == destination_table.to_column_2 AND
+        origin_table.from_column_3 == destination_table.to_column_3
+
+    The type of join (inner vs left, etc.) is not specified.
+    The tables are not specified.
+    """
+
+    # (from_column, to_column) pairs, where from_column is on the origin table
+    # and to_column is on the destination table of the join.
+    column_pairs: Set[Tuple[str, str]]
```

Comment:
Consider `FrozenSet` or `AbstractSet` instead here. `Set` is literally `set()`, the explicitly-mutable kind. `AbstractSet` is sufficient to make it immutable-in-practice because abstract sets don't have mutability functions defined and `mypy` will prevent mutations in typed code unless you literally `isinstance` or `cast` to a mutable set.